streaming/aggregate/first.go

```diff
@@ -3,23 +3,106 @@ package aggregate
 import (
 	"context"
 
+	"github.com/pkg/errors"
+
 	"github.com/cube2222/octosql"
 	"github.com/cube2222/octosql/streaming/storage"
 )
 
+var currentFirstPrefix = []byte("$current_first$")
+var currentFirstCountsPrefix = []byte("$current_first_counts$")
+
 type First struct {
 }
 
+func NewFirstAggregate() *First {
+	return &First{}
+}
+
 func (agg *First) AddValue(ctx context.Context, tx storage.StateTransaction, value octosql.Value) error {
-	panic("not implemented yet")
+	currentFirstStorage := storage.NewDeque(tx.WithPrefix(currentFirstPrefix))
+	currentFirstCountsStorage := storage.NewMap(tx.WithPrefix(currentFirstCountsPrefix))
+
+	var currentValueCount octosql.Value
+	err := currentFirstCountsStorage.Get(&value, &currentValueCount)
+	if err == storage.ErrNotFound {
+		currentValueCount = octosql.MakeInt(0)
+	} else if err != nil {
+		return errors.Wrap(err, "couldn't get current value count from first storage")
+	}
+
+	currentValueCount = octosql.MakeInt(currentValueCount.AsInt() + 1)
+
+	err = currentFirstStorage.PushBack(&value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't push back current value to first deque")
+	}
+
+	err = currentFirstCountsStorage.Set(&value, &currentValueCount)
+	if err != nil {
+		return errors.Wrap(err, "couldn't set current value count in first storage")
+	}
+
+	return nil
 }
 
 func (agg *First) RetractValue(ctx context.Context, tx storage.StateTransaction, value octosql.Value) error {
-	panic("not implemented yet")
+	currentFirstCountsStorage := storage.NewMap(tx.WithPrefix(currentFirstCountsPrefix))
+
+	var currentValueCount octosql.Value
+	err := currentFirstCountsStorage.Get(&value, &currentValueCount)
+	if err == storage.ErrNotFound {
+		currentValueCount = octosql.MakeInt(0)
+	} else if err != nil {
+		return errors.Wrap(err, "couldn't get current value count from first storage")
+	}
+
+	currentValueCount = octosql.MakeInt(currentValueCount.AsInt() - 1)
+
+	if currentValueCount.AsInt() == 0 { // current value was just cleared, no need to store its count or retractions count
+		err = currentFirstCountsStorage.Delete(&value)
+		if err != nil {
+			return errors.Wrap(err, "couldn't delete current value from first storage")
+		}
+	} else {
+		err = currentFirstCountsStorage.Set(&value, &currentValueCount)
+		if err != nil {
+			return errors.Wrap(err, "couldn't set current value count in first storage")
+		}
+	}
+
+	return nil
 }
 
 func (agg *First) GetValue(ctx context.Context, tx storage.StateTransaction) (octosql.Value, error) {
```

Comment:
The approach with keeping count is a very nice idea, and it's intuitive once one understands what's happening (it becomes clear in GetValue). A comment in AddValue/RetractValue would clear things up ^^